Improve UCX documentation and examples (#545)
rapidsai/dask-cuda: docs/source/ucx.rst

Typically, these will affect ``UCX_TLS`` and ``UCX_SOCKADDR_TLS_PRIORITY``, environment variables used by UCX to decide what transport methods to use and which to prioritize, respectively.
However, some will affect related libraries, such as RMM:

- ``ucx.cuda_copy: true`` -- **required.**

  Adds ``cuda_copy`` to ``UCX_TLS``, enabling CUDA transfers over UCX.

- ``ucx.tcp: true`` -- **required.**

  Adds ``tcp`` to ``UCX_TLS``, enabling TCP transfers over UCX; this is required for very small transfers which are inefficient for NVLink and InfiniBand.

- ``ucx.nvlink: true`` -- **required for NVLink.**

  Adds ``cuda_ipc`` to ``UCX_TLS``, enabling NVLink transfers over UCX; affects intra-node communication only.

- ``ucx.infiniband: true`` -- **required for InfiniBand.**

  Adds ``rc`` to ``UCX_TLS``, enabling InfiniBand transfers over UCX.

- ``ucx.rdmacm: true`` -- **recommended for InfiniBand.**

  Replaces ``sockcm`` with ``rdmacm`` in ``UCX_TLS`` and ``UCX_SOCKADDR_TLS_PRIORITY``, enabling remote direct memory access (RDMA) for InfiniBand transfers.

>   This is recommended by UCX for use with InfiniBand, and will not work it is disabled.

Reviewer: ```suggestion
  This is recommended by UCX for use with InfiniBand, and will not work if InfiniBand is disabled.
```

- ``ucx.net-devices: <str>`` -- **recommended.**

  Explicitly sets ``UCX_NET_DEVICES`` instead of defaulting to ``"all"``, which can result in suboptimal performance.
  If using InfiniBand, set to ``"auto"`` to automatically detect the InfiniBand interface closest to each GPU.
  If InfiniBand is disabled, set to a UCX-compatible ethernet interface, e.g. ``"enp1s0f0"`` on a DGX-1.
  All available UCX-compatible interfaces can be listed by running ``ucx_info -d``.

  .. warning::
      Setting ``ucx.net-devices: "auto"`` assumes that all InfiniBand interfaces on the system are connected and properly configured; undefined behavior may occur otherwise.
  

- ``rmm.pool-size: <str|int>`` -- **recommended.**

  Allocates an RMM pool of the specified size for the process; size can be provided with an integer number of bytes or in human readable format, e.g. ``"4GB"``.
  It is recommended to set the pool size to at least the minimum amount of memory used by the process; if possible, one can map all GPU memory to a single pool, to be utilized for the lifetime of the process.

.. note::
    These options can be used with mainline Dask/Distributed.

>     However, this will disable a variety of features, such as the automatic detection of InfiniBand interfaces. 

Reviewer: ```suggestion
    However, some features are exclusive to Dask-CUDA, such as the automatic detection of InfiniBand interfaces. 
```

    See :doc:`Specializations for GPU Usage <specializations>` for more details on the benefits of using Dask-CUDA.

Usage
-----

Dask-CUDA workers using UCX communication can be started manually with the ``dask-cuda-worker`` CLI tool or automatically with ``LocalCUDACluster``.
In either case, a ``distributed.Client`` must be made for the worker cluster using the same UCX configuration.

dask-cuda-worker
^^^^^^^^^^^^^^^^

A Dask cluster with UCX support can be started using the ``dask-cuda-worker`` CLI tool with a Dask scheduler which has been configured for UCX.

> *Do we want to include any reasons you would do this instead of using LocalCUDACluster?*

Reviewer: `LocalCUDACluster` is single-node only, anything else requires `dask-cuda-worker`.

Scheduler
"""""""""

UCX configuration options will need to be specified for ``dask-scheduler`` as environment variables; see `Dask Configuration <https://docs.dask.org/en/latest/configuration.html#environment-variables>`_ for more details on the mapping between environment variables and options.

To start a Dask scheduler using UCX with all supported transports and a 1 gigabyte RMM pool:

.. code-block:: bash

    DASK_UCX__CUDA_COPY=True \
    DASK_UCX__TCP=True \
    DASK_UCX__NVLINK=True \
    DASK_UCX__INFINIBAND=True \
    DASK_UCX__RDMACM=True \
    DASK_UCX__NET_DEVICES=mlx5_0:1 \
    DASK_RMM__POOL_SIZE=1GB \
    dask-scheduler --protocol ucx --interface ib0

> Note the specification of ``mlx5_0:1`` as our UCX net device; because the scheduler does not rely upon Dask-CUDA, it cannot automatically detect InfiniBand interfaces, so we must specify one explicitly *(is this true?)*.

Reviewer: Yes, it is true.

We communicate to the scheduler that we will be using UCX with the ``--protocol`` option, and that we will be using InfiniBand with the ``--interface`` option.

To start the same Dask scheduler as above but only using NVLink:

.. code-block:: bash

    DASK_UCX__CUDA_COPY=True \
    DASK_UCX__TCP=True \
    DASK_UCX__NVLINK=True \
    DASK_RMM__POOL_SIZE=1GB \
    dask-scheduler --protocol ucx --interface eth0

Note that we no longer specify a net device, as this generally can be skipped when using a non-InfiniBand interface.

Workers
"""""""

All the relevant Dask configuration options for UCX have analogous parameters in ``dask-cuda-worker``; see :doc:`Worker <worker>` for a complete list of these options.

> To start workers with all supported transports and a 1 gigabyte RMM pool *(do we need to specify an interface here?)*:

Reviewer: In most cases, no, UCX will be capable of identifying the optimal network interface to use.

        ucx_net_devices="auto",
        rmm_pool_size="1GB"
    )

Client
^^^^^^

The UCX configurations used by the scheduler and client should be the same.
This can be ensured by using ``dask_cuda.initialize``, a utility which takes the same UCX configuring arguments as ``LocalCUDACluster`` and adds them to the current Dask configuration used when creating the scheduler and client; see the :doc:`API reference <api>` for a complete list of arguments.

To start a cluster and client with all supported transports and a 1 gigabyte RMM pool:

.. code-block:: python

    from dask.distributed import Client
    from dask_cuda import LocalCUDACluster
    from dask_cuda.initialize import initialize

    initialize(
        enable_tcp_over_ucx=True,
        enable_nvlink=True,
        enable_infiniband=True,
        enable_rdmacm=True,
        net_devices="mlx5_0:1",
    )
    cluster = LocalCUDACluster(
        protocol="ucx",
        interface="ib0", # passed to the scheduler
        enable_tcp_over_ucx=True,
        enable_nvlink=True,
        enable_infiniband=True,
        ucx_net_devices="auto",
        rmm_pool_size="1GB",
    )
    client = Client(cluster)

> Note the specification of ``"mlx5_0:1"`` as our net device in ``initialize()``; because the scheduler and client do not rely upon Dask-CUDA, they cannot automatically detect InfiniBand interfaces, so we must specify one explicitly (is this true?).

Reviewer: Yes, as in the previous comment in the scheduler section.

UCX Integration
===============

Communication can be a major bottleneck in distributed systems.
Dask-CUDA addresses this by supporting integration with `UCX <https://www.openucx.org/>`_, an optimized communication framework that provides high-performance networking and supports a variety of transport methods, including:

- NVLink for *something ubiquitous to NVLink systems*

> - InfiniBand for *something ubiquitous to InfiniBand systems*

Author: Sure! More details would probably be better suited for a hardware requirements section.

- TCP for systems that do not have specialized hardware

This integration is enabled through `UCX-Py <https://ucx-py.readthedocs.io/>`_, an interface that provides Python bindings for UCX.

Requirements
------------

Hardware
^^^^^^^^

*Go into more detail on what hardware is required for NVLink, InfiniBand support*

Software
^^^^^^^^

> *Does UCX integration require anything else other than UCX-Py?*

Author: Thanks for the link! I will also go into some detail about when you would want to use the Conda package vs building/installing from source (essentially whether or not you plan to use IB/RDMA).

> *Does UCX integration require anything else other than UCX-Py?*

Reviewer: I think that information is better suited to the UCX-Py docs link I mentioned above. Let's try to avoid duplicating information which will eventually become inconsistent between the two docs.

When using UCX, each NVLink and InfiniBand memory buffer must create a mapping between each unique pair of processes they are transferred across; this can be quite costly, taking up to 100 ms per mapping.
For this reason, it is strongly recommended to use `RAPIDS Memory Manager (RMM) <https://github.com/rapidsai/rmm>`_ to allocate a memory pool that is only prone to a single mapping operation, which all subsequent transfers may rely upon.

Configuration
^^^^^^^^^^^^^

In addition to installations of UCX and UCX-Py on your system, several options must be specified within your Dask configuration to enable the integration.
Typically, these will affect ``UCX_TLS`` and ``UCX_SOCKADDR_TLS_PRIORITY``, environment variables used by UCX to decide what transport methods to use and which to prioritize, respectively.
However, some will affect related libraries, such as RMM:

- ``ucx.cuda_copy: true`` -- **required.**
    
    Adds ``cuda_copy`` to ``UCX_TLS``, enabling *all* transfers over UCX.
    *Is this accurate to say?*

- ``ucx.tcp: true`` -- **required.**

    Adds ``tcp`` to ``UCX_TLS``, enabling TCP transfers over UCX if NVLink or InfiniBand are unavailable or disabled.

>     *Why is this required if they are available and enabled?*

Author: Would it maybe be better to use that for the reasoning to enable it in this case?

>     *Why is this required if they are available and enabled?*

Reviewer: You can write a brief comment, but we probably don't want to rant about it here for too long as it's hard to draw a line of how much detail we want to go in. Here again a long rant would probably suit better UCX-Py (not suggesting we need to do it).

- ``ucx.nvlink: true`` -- required for NVLink.

    Adds ``cuda_ipc`` to ``UCX_TLS``, enabling NVLink transfers over UCX; affects intra-node communication only.

- ``ucx.infiniband: true`` -- required for InfiniBand.

    Adds ``rc`` to ``UCX_TLS``, enabling InfiniBand transfers over UCX; affects inter-node communication only.

- ``ucx.rdmacm: true`` -- recommended for InfiniBand.

    Replaces ``sockcm`` with ``rdmacm`` in ``UCX_TLS`` and ``UCX_SOCKADDR_TLS_PRIORITY``, *enabling remote direct memory access (RDMA) for connection management.*

>     This is recommended by UCX for use with InfiniBand, and will have no effect if InfiniBand tranfers are disabled.

Author: Is that to say that unexpected behavior will occur if `ucx.rdmacm` is enabled but `ucx.infiniband` is not?

>     This is recommended by UCX for use with InfiniBand, and will have no effect if InfiniBand tranfers are disabled.

Reviewer: Not unexpected behavior, the behavior is very much expected, resulting in an error on initialization.

- ``ucx.net-devices: <str>`` -- recommended.

    Explicitly sets ``UCX_NET_DEVICES`` instead of defaulting to ``"all"``, which can result in suboptimal performance.
    If using InfiniBand, set to the desired IB device, e.g. ``"mlx5_0:1"``.
    If InfiniBand is disabled, set to the ethernet device, e.g. ``"enp1s0f0"`` on a DGX-1.

>     All available UCX-compatible devices can be listed by running ``ucx_info -d`` or ``ifconfig`` *(are all the options under ifconfig compatible)*.

Reviewer: I don't think all options under `ifconfig` are valid, for example `veth*` (often used by Docker) aren't. `ucx_info -d` only should be more reliable.

Also, interestingly the most important use case for us is missing in this section (it was missing before your proposed changes here too), which is the `"auto"` option, as documented in https://github.com/rapidsai/dask-cuda/blob/e72d776645ee3ee7e1672f73c9c043731cf401d5/dask_cuda/cli/dask_cuda_worker.py#L187-L203 and https://github.com/rapidsai/dask-cuda/blob/e72d776645ee3ee7e1672f73c9c043731cf401d5/dask_cuda/local_cuda_cluster.py#L72-L86 . Could you add that here as well? It's documented in "Important Notes" in this file below, but it seems sensible to have it here too.

Author: It does make sense to include information about `"auto"` here. Are there any resources on installing UCX-Py with hwloc support? Didn't see anything explicit about it in the installation guide.

Reviewer: It's in the build dependencies: https://ucx-py.readthedocs.io/en/latest/install.html#build-dependencies

Author: Question about specifying an interface when InfiniBand is disabled:

I notice later on in the documentation, we say when setting up a scheduler for UCX that setting `DASK_UCX__NET_DEVICES` can generally be skipped when using a non-InfiniBand interface is used. Is this universally true, or does it only apply to the scheduler?

Reviewer: This can apply to all processes. `UCX_NET_DEVICES` is merely a way to limit the visibility of network devices to UCX, meaning it won't "see" any others. We can still choose the device by specifying `--interface` to scheduler and workers, which is generally preferred if no IB is used. To be even more explicit, we're likely going to drop `DASK_UCX__NET_DEVICES` entirely starting from UCX 1.10, because UCX is now capable of ensuring it uses the IB interface topologically closest to the GPU (which is what `"auto"` does now).

Author: In this case, should I remove the recommendation to explicitly set `ucx.net-devices` to an ethernet device when InfiniBand is disabled? Seems like an inconsistency, unless I'm missing something.

Reviewer: The way I read is we recommend users to set to the "optimal" device, so I would still leave it in place. I know in the past we had issues with suboptimal devices being chosen and it seems like UCX 1.10+ is better in that regard, so what I meant to say was that starting from UCX 1.10 we'll then prefer not to set `UCX_NET_DEVICES`, but I don't think in our current state (UCX 1.9), we're ready for that.

- ``rmm.pool-size: <str | int>`` -- recommended.

    Allocates an RMM pool of the specified size for the process; size can be provided with an integer number of bytes or in human readable format, e.g. ``"4GB"``.
    In addition to reducing the cost of mapping incurred by memory transfers, a pool can prevent the Dask scheduler from deserializing CUDA data and causing a crash.
    It is recommended to set the pool size to at least the minimum amount of memory used by the process; if possible, one can map all GPU memory to a single pool, to be utilized for the lifetime of the process.

.. note::
    These options can also be used with mainline Dask/Distributed.

>     However, this may disable a variety of features, such as *put some features exclusive to Dask-CUDA here*. 

Reviewer: `net-devices="auto"` is an exclusive feature of Dask-CUDA, I don't think there are others that apply to this section. The line below this should also be clarifying in this context.

Author: Yeah, so far from the config section alone I have:

- Use of CUDA transfers for task on GPU
- Setting `net-devices="auto"`

Anything I'm missing here?

Reviewer: > Use of CUDA transfers for task on GPU

This is again not true. A user can very much use mainline Dask with GPUs, they will only not benefit from the specializations Dask-CUDA provides -- thus my comment on the line that is immediately below this clarifying that difference. Therefore, the only thing I can think of here is indeed `net-devices="auto"`.
